File: backend/src/main/java/com/hotel/reservation/service/RoomService.java
package com.hotel.reservation.service;

import com.hotel.reservation.dto.RoomDTO;
import com.hotel.reservation.dto.RoomSearchCriteria;
import java.time.LocalDate;
import java.util.List;

/**
 * Service interface for Room operations.
 */
public interface RoomService {
    RoomDTO createRoom(RoomDTO roomDTO);
    RoomDTO updateRoom(Long id, RoomDTO roomDTO);
    void deleteRoom(Long id);
    RoomDTO getRoomById(Long id);
    List<RoomDTO> getAllRooms();
    List<RoomDTO> searchRooms(RoomSearchCriteria criteria);
    boolean checkAvailability(Long roomId, LocalDate checkInDate, LocalDate checkOutDate);
}

File: frontend/src/services/roomService.ts
import apiClient from './apiService';
import { Room, RoomSearchCriteria } from '../utils/types';

export const roomService = {
  async getAllRooms(): Promise<Room[]> {
    const response = await apiClient.get<Room[]>('/rooms');
    return response.data;
  },

  async getRoomById(id: number): Promise<Room> {
    const response = await apiClient.get<Room>(`/rooms/${id}`);
    return response.data;
  },

  async searchRooms(criteria: RoomSearchCriteria): Promise<Room[]> {
    const response = await apiClient.post<Room[]>('/rooms/search', criteria);
    return response.data;
  },

  async getAvailableRooms(checkIn: string, checkOut: string): Promise<Room[]> {
    const response = await apiClient.get<Room[]>('/rooms/available', {
      params: { checkIn, checkOut }
    });
    return response.data;
  },

  async createRoom(room: Omit<Room, 'id'>): Promise<Room> {
    const response = await apiClient.post<Room>('/rooms', room);
    return response.data;
  },

  async updateRoom(id: number, room: Partial<Room>): Promise<Room> {
    const response = await apiClient.put<Room>(`/rooms/${id}`, room);
    return response.data;
  },

  async deleteRoom(id: number): Promise<void> {
    await apiClient.delete(`/rooms/${id}`);
  },

  async updateAvailability(id: number, available: boolean): Promise<Room> {
    const response = await apiClient.patch<Room>(`/rooms/${id}/availability`, null, {
      params: { available }
    });
    return response.data;
  }
};
